velox/docs/develop/hash-table.rst

```diff
@@ -0,0 +1,117 @@
+
+.. image:: images/hash-table-insert1.png
+    :align: center
+
+In this case, we simply insert our entry into the first slot of the bucket.
+We store only the tag (7 bits of the hash value) in the hash table. The hash number itself is not stored.
+The pointer refers to memory outside of the HashTable where the complete value is stored.
+This is usually a row in a RowContainer. The HashTable can be viewed as an index on top of a RowContainer
+that helps locate records with matching keys faster. HashTable itself doesn’t store data or keys.
+
+.. image:: images/hash-table-insert2.png
+    :align: center
+
+Another possibility is a bucket partially filled (for example, has one slot occupied like above).
+In this case there is a possibility that the new entry is a duplicate of the one already stored.
+Hence, we compare the tag of the new entry with tags stored in the bucket.
+If none match, this entry is not a duplicate, so we store it in the next available slot in the bucket.
+
+.. image:: images/hash-table-insert3.png
+    :align: center
+
+However, if one or more existing tags match the tag of the new entry, we follow the pointers to compare
+the keys to determine whether there is a match or not. If there is no match we insert a new entry.
+Otherwise, we drop it.
+
+A final possibility is that the bucket is full.
+
+.. image:: images/hash-table-insert3.png
+    :align: center
+
+First, we need to check if the new entry is a duplicate of one of the 16 entries stored in the bucket.
+We compare tags and if necessary follow the pointers to compare the keys.
+If there is a match, we drop the entry and we are done. If there is no match, we go to the
+next bucket and repeat the process. In rare cases, we may end up checking many buckets until we find a
+duplicate existing entry or an empty slot for the new entry. That’s why it is important to ensure that
+the Hash Table is never full and there are enough gaps in the form of empty slots.
+
+Resizing
+--------
+
+If HashTable fills up beyond ⅞ of capacity, it needs to be resized. Each resize doubles the capacity.
+We allocate a new HashTable and insert all existing entries using the “Inserting an entry” process.
+Since we know that all entries are unique, the “Inserting an entry” process can be simplified to
+eliminate the logic for checking whether a new entry is a duplicate of an existing one. Hence, to
+insert an entry, we compute a hash, extract tag and bucket number, go to the bucket and insert the
+entry if there is space. If the bucket is full, we proceed to the next bucket and continue until we
+find a bucket with an empty slot. We insert the new entry there.
+
+Use Cases
+---------
+
+The main use cases for the HashTable are `Join <joins.html>`_ and
+`Aggregation <aggregations.html>`_ operators.
+
+HashBuild operator builds the HashTable to store unique values of the join keys found on the build
```

Comment:
In HashJoin we do allow multiple rows with the same key on the build side though. So is that managed through some separate chaining logic ?

Reply:
Replied earlier.